docs/design/interpreter/compiled-code-interop.md

```diff
@@ -0,0 +1,69 @@
+
+##### Interpreter entry
+
+Given we can't dynamically generate thunks that can be invoked, in order to generate an interpreter entry point we could reuse the functionality of fat pointers that is already used with native aot. Fat pointers can point to additional data, if a bit is set, rather than being actual function pointers. Instead of calling the pointer directly, the calling code will check for the most significant bit. If it is set, it will instead dereference the pointer and obtain the real function pointer together with the additional argument that is passed to the call. For the purpose of entering the interpreter, we would generate a fat pointer that has the target destination as a compiled IL wrapper for the signature in question together with the interpreter method pointer that is passed. The wrapper will obtain the pointer to the interpreter stack, will move all arguments there and call into the C++ interpreter path, passing the method and the address on the stack where the arguments have been written. When aot compiling an assembly, we would need to consider every single call as a potential entry to the interpreter and, if we deem it possible, we would generate an interpreter entry wrapper for the call signature in question.
+
+Native code has no knowledge of fat pointers, so we will need to explicitly generate small thunks for every single `UnmanagedCallersOnly` method. On mono, during app compilation, a build task scans the assembly for all `UnmanagedCallersOnly` methods and it dynamically generates a separate C method for each one of them. This method will have its own data that is later initialized with the function pointer for the cconv translation wrapper together with the interpreter method pointer argument. Rather than have this logic in special build tasks that dynamically generate C code, it might make more sense to  simply generate a special direct call wrapper for `UnmanagedCallersOnly` methods, that are not being aot compiled to the wasm image (this might represent a scenario that is just good to have, for interp-only, but not really mandatory since we can choose to always aot compile these methods).
```

Comment:
> this might represent a scenario that is just good to have, for interp-only, but not really mandatory since we can choose to always aot compile these methods

Right. I think we should not consider this a need. It is good to have an idea on how to address it, but I think we should avoid any "nice to have" scenarios for the interpreter.

Reply:
This is best to have for the purpose of incrementally developing the interpreter by using an interp-only configuration, in case certain parts of the compiled code interop lag behind, since they might be more challenging. CoreCLR with aot wasm doesn't even exist at the moment and I have no idea how easy is to reuse stuff from nativeaot. For wasm seems very likely to me that we would have an interp-only first.